### PR TITLE
API for framework

### PR DESCRIPTION
**Pull request by prince-7**
Description:
* maryam-web file is included for now so that testing can be performed easily(would be removed later)
* Current API result:-
command:- 
```
curl -L -X POST 'http://localhost:5000/api/modules?module=github&query=owasp&limit=1&output=true'
```
Result:
```
{
  "meta": {
    "workspace": "default",
    "cmd": "github -q owasp -l 1 --output --output",
    "module": {
      "author": "Aman Singh",
      "description": "Search your query in the GitHub and show the results.",
      "examples": [
        "github -q <QUERY> -l 15 -e carrot2,bing,qwant --output"
      ],
      "name": "GitHub Search",
      "options": [
        [
          "query",
          null,
          true,
          "Query string",
          "-q",
          "store",
          "str"
        ],
        [
          "limit",
          1,
          false,
          "Search limit(number of pages, default=1)",
          "-l",
          "store",
          "int"
        ],
        [
          "count",
          50,
          false,
          "Number of links per page(min=10, max=100, default=50)",
          "-c",
          "store",
          "int"
        ],
        [
          "engine",
          "google,github",
          false,
          "Engine names for search(default=google, github)",
          "-e",
          "store",
          "str"
        ],
        [
          "thread",
          2,
          false,
          "The number of engine that run per round(default=2)",
          "-t",
          "store",
          "int"
        ],
        [
          "output",
          false,
          false,
          "Save the output to the workspace",
          "--output",
          "store_true",
          "bool"
        ],
        [
          "api",
          false,
          false,
          "Show results in the JSON format",
          "--api",
          "store_true",
          "bool"
        ],
        [
          "format",
          false,
          false,
          "Beautifying JSON output if --api is used",
          "--format",
          "store_true",
          "bool"
        ],
        [
          "output",
          false,
          false,
          "Save the output to the workspace",
          "--output",
          "store_true",
          "bool"
        ],
        [
          "api",
          false,
          false,
          "Show results in the JSON format",
          "--api",
          "store_true",
          "bool"
        ],
        [
          "format",
          false,
          false,
          "Beautifying JSON output if --api is used",
          "--format",
          "store_true",
          "bool"
        ]
      ],
      "sources": [
        "google",
        "carrot2",
        "bing",
        "yippy",
        "yahoo",
        "millionshort",
        "qwant",
        "duckduckgo",
        "github"
      ],
      "version": "1.0"
    }
  },
  "output": {
    "repositories": [
      "https://github.com/wisec/OWASP-Testing-Guide-v5",
      "https://github.com/tanprathan/OWASP-Testing-Checklist",
      "https://github.com/OWASP/DevGuide",
      "https://github.com/OWASP/railsgoat",
      "https://github.com/OWASP/IoTGoat",
      "https://github.com/OWASP/Top10",
      "https://github.com/OWASP/CheatSheetSeries",
      "https://github.com/OWASP/ASVS",
      "https://github.com/OWASP/Go-SCP",
      "https://github.com/chuckfw/owaspbwa",
      "https://github.com/OWASP/Serverless-Top-10-Project",
      "https://github.com/OWASP/Nettacker",
      "https://github.com/OWASP/SafeNuGet",
      "https://github.com/OWASP/owasp-masvs",
      "https://github.com/OWASP/NodeGoat",
      "https://github.com/OWASP/IoT-Security-Verification-Standard-ISVS",
      "https://github.com/0xRadi/OWASP-Web-Checklist",
      "https://github.com/OWASP/java-html-sanitizer",
      "https://github.com/SasanLabs/VulnerableApp",
      "https://github.com/owasp-cloud-security/owasp-cloud-security",
      "https://github.com/OWASP/wstg",
      "https://github.com/bkimminich/juice-shop",
      "https://github.com/OWASP/www-community",
      "https://github.com/OWASP/owasp-mstg",
      "https://github.com/SpiderLabs/owasp-modsecurity-crs",
      "https://github.com/OWASP/igoat",
      "https://github.com/zaproxy/zaproxy",
      "https://github.com/OWASP/OWASP-VWAD",
      "https://github.com/jeremylong/DependencyCheck",
      "https://github.com/OWASP/SecurityShepherd",
      "https://github.com/OWASP/Benchmark",
      "https://github.com/OWASP/ZSC",
      "https://github.com/OWASP/glue",
      "https://github.com/OWASP/OWASP-WebScarab",
      "https://github.com/OWASP/SecureTea-Project",
      "https://github.com/OWASP/Amass",
      "https://github.com/OWASP/samm",
      "https://github.com/OWASP/API-Security",
      "https://github.com/OWASP/joomscan",
      "https://github.com/OWASP/www-chapter-london",
      "https://github.com/acrcdlsd/OWASP",
      "https://github.com/OWASP/PureCaptcha",
      "https://github.com/tjunxiang92/Android-Vulnerabilities",
      "https://github.com/NoobieDog/Dir-Xcan",
      "https://github.com/zaproxy/zaproxy-website",
      "https://github.com/1N3/BlackWidow",
      "https://github.com/thesp0nge/owasp-esapi-ruby",
      "https://github.com/zaproxy/zap-api-dotnet",
      "https://github.com/rapPayne/WebGoat.Net"
    ],
    "blogs": [
      "spiderlabs.github.com",
      "owasp.github.com"
    ],
    "usernames": [
      "SasanLabs",
      "owasp-cloud-security",
      "0xRadi",
      "bkimminich",
      "tanprathan",
      "chuckfw",
      "wisec",
      "SpiderLabs",
      "zaproxy",
      "owasp-modsecurity-crs",
      "jeremylong",
      "owasp",
      "OWASP",
      "psiinon",
      "Ali-Razmjoo",
      "rezasp",
      "OWASPFoundation",
      "cldrn",
      "ManhNho",
      "sid22",
      "Priya997",
      "caffix"
    ]
  }
}
```

### CODE REVIEW

File: requirements [hunk 1]
```diff
@@ -1,3 +1,6 @@
-requests
-cloudscraper
-bs4
\ No newline at end of file
+requests
+cloudscraper
+bs4
+flask
```
Comment (saeeddhqan) [inline requirements:5]: Do we need all of them?

Comment (prince-7) [reply requirements:5]: yes

**Comment by saeeddhqan** [comment]
Solve the conflicts with new version

**Comment by prince-7** [comment]
Conflicts Resolved

**Comment by saeeddhqan** [comment]
cmd
it looks to load the core twice. why?

**Comment by saeeddhqan** [comment]
cmderror
Why does it show the errors in HTML format?

**Comment by prince-7** [comment]
>it looks to load the core twice. why?

once in init.py and once in api.py

>Why does it show the errors in HTML format?

Because the correct way to  make the request would be
```
curl -L -X POST 'http://localhost:5000/api/modules?module=github&query=owasp&thread=2&limit=1&output=true'
```

**Comment by saeeddhqan** [comment]
The question is Why does it show the errors in HTML format? not JSON?

**Comment by saeeddhqan** [comment]
Also, my query doesn't have any problem! maybe the user wants to search for something like that!

**Comment by prince-7** [comment]
>The question is Why does it show the errors in HTML format? not JSON?

**Comment by prince-7** [comment]
>The question is Why does it show the errors in HTML format? not JSON?

Because you have used ' ' (space) which should be %20 in URL encoding

>Also, my query doesn't have any problem! maybe the user wants to search for something like that!

so search this way
```
curl -L -X POST 'http://localhost:5000/api/modules?module=github&query=bitso%20-t%202&limit=1&output=true'
```

**Comment by saeeddhqan** [comment]
get is better than post. don't show the modules and their information by default. the user can do it with /framework endpoint.
also http://localhost:5000/api/ is blank.

Isn't it better to use the socket instead of the flask or switch it to Golang? Golang is very fast and simple for API!

**Comment by prince-7** [comment]
>also http://localhost:5000/api/ is blank.

That can be written later, that will have definitions of requests and endpoints

>Isn't it better to use the socket instead of the flask or switch it to Golang? 

The whole project is in python that's why using flask is the right choice.

>Golang is very fast and simple for API!

yes it is but later this will be a complete web interface that will have a frontend, so therefore flask.

**Comment by prince-7** [comment]
>get is better than post. don't show the modules and their information by default. the user can do it with /framework endpoint.

Now /api/framework/  - > get request -> shows module data
 /api/modules/ -> get request -> runs modules

**Comment by saeeddhqan** [comment]
Update the pr with API branch (don't apply iris changes. last commit).

**Comment by prince-7** [comment]
Why close this?

**Pull request closed without merging**